dali/python/nvidia/dali/worker.py

```diff
@@ -0,0 +1,233 @@
+    to avoid blocking of the main process when it schedules another batch for the worker and
+    worker is busy with previously scheduled computations.
+    """
+    try:
+        while True:
+            scheduled = task_pipe.recv()
+            if scheduled is None:
+                break
+            with tasks_cv:
+                tasks_queue.append(scheduled)
+                tasks_cv.notify()
+    finally:
+        with tasks_cv:
+            tasks_queue.insert(0, None)
+            tasks_cv.notify()
+
+
+def worker(worker_id, callbacks, prefetch_queue_depths, initial_chunk_size, task_pipe, res_pipe, sock):
+    """Entry point of worker process.
+
+    Computes the data in the main thread, in separate threads:
+    * waits for incoming tasks,
+    * serializes results and passes them to the main process.
+
+    Parameters
+    ----------
+    `callbacks` : callable list
+        List of callables that worker can call to perform a (part of parallelized) task.
+    `prefetch_queue_depths` : int
+        Number of shared memory chunks that should be allocated per callaback, used in cycle buffer manner
+        to pass callback results to parent process.
+    `initial_chunk_size` : int
+        Initial size of shared memory chunk.
+    `task_pipe`: Pipe
+        Pipe used to read list of tasks that given callback should be run on to produce (part of a) result batch.
+    `res_pipe`: Pipe
+        Pipe used to notify the parent process about another batch ready to read in the given memory chunk.
+    `sock` : socket
+        Python wrapper around Unix socket used to pass file descriptors identifying shared memory chunk to parent process.
+    """
+    contexts = None
+    ready_cv = threading.Condition()
+    tasks_cv = threading.Condition()
+    ready_queue, tasks_queue = [], []
+    batch_dispatcher = SharedBatchesDispatcher(worker_id, sock, res_pipe)
+    # run the thread as a daemon so that even when results queue blocks worker process can exit anyway
+    # and can be joined in the parent process
+    dispatcher_thread = threading.Thread(target=dispatcher, args=(
+        batch_dispatcher, ready_cv, ready_queue), daemon=True)
+    receiver_thread = threading.Thread(target=receiver, args=(
+        task_pipe, tasks_cv, tasks_queue), daemon=True)
+    dispatcher_thread.start()
+    receiver_thread.start()
+    try:
+        contexts = [
+            CallbackContext(callback, [
+                SharedMemChunk("chunk_{}_{}_{}".format(worker_id, callback_idx, prefetch_idx), initial_chunk_size)
+                for prefetch_idx in range(prefetch_queue_depth)
+            ])
+            for callback_idx, (callback, prefetch_queue_depth) in enumerate(zip(callbacks, prefetch_queue_depths))
+        ]
+        while True:
+            with tasks_cv:
+                while len(tasks_queue) == 0:
+                    tasks_cv.wait()
+                scheduled = tasks_queue.pop(0)
```

Comment:
Scheduled? I think this task is no longer just scheduled - it's headed for immediate execution, isn't it?
```suggestion
                task = tasks_queue.pop(0)
```